### DISCUSSION TITLE
Transitions between AutoDJ and Live (switch to LIVE only after the current AutoDJ song ends)

### DISCUSSION THREAD

**Discussion by unudelaradio** [savonet/liquidsoap#4011] [Q&A]
Hello!

I've playing around these days with custom Liquidsoap configurations in an older Azuracast version and i discovered these GREAT transitions:

- streamer connects, but the switch is happening only after the current AutoDJ song finishes
- on transition: AutoDJ _fades out_ and Live _fades in_ same time (mixing together)
- when Live ends there's also an awesome overlap: Live is still playing and while is fading out, the next song from AutoDJ overlaps it for few seconds without fade in (because the song has custom fade in set to 0).

This is the most wonderful scenario! And i think is possible.... or at least it was in older versions (Azuracast v0.17.7 and i guess Liquidsoap version is older than 2.2.x).

But i don't have any coding knowledge so i'm asking for your help on how improve some things. My goal is to include Stereo Tool processing, but only for AutoDJ, not for Live. 

Could someone help me with a custom LS code for that, but keeping the exact behavior described above?

Here's my code for that awesome transitions:

```
# DJ Authentication
last_authenticated_dj = ref("")
live_dj = ref("")

def dj_auth(login) =
    auth_info =
        if (login.user == "source" or login.user == "") and (string.match(pattern="(:|,)+", login.password)) then
            auth_string = string.split(separator="(:|,)", login.password)
            {user = list.nth(default="", auth_string, 0),
            password = list.nth(default="", auth_string, 2)}
        else
            {user = login.user, password = login.password}
        end
    
    response = azuracast_api_call(
        timeout_ms=5000,
        "auth",
        json.stringify(auth_info)
    )
    
    if (response == "true") then
        last_authenticated_dj := auth_info.user
        true
    else
        false
    end
end

def live_connected(header) =
    dj = !last_authenticated_dj
    log("DJ Source connected! Last authenticated DJ: #{dj} - #{header}")

    live_enabled := true
    live_dj := dj
    
    _ = azuracast_api_call(
        timeout_ms=5000,
        "djon",
        json.stringify({user = dj})
    )
end

def live_disconnected() =
    _ = azuracast_api_call(
        timeout_ms=5000,
        "djoff",
        json.stringify({user = !live_dj})
    )
    
    live_enabled := false
    live_dj := ""
end

# A Pre-DJ source of radio that can be broadcast if needed',
radio_without_live = radio
ignore(radio_without_live)

# Live Broadcasting
live = input.harbor("/", port = 8005, password="*******", icy = true, icy_metadata_charset = "UTF-8", metadata_charset = "UTF-8", on_connect = live_connected, on_disconnect = live_disconnected, buffer = 10.00, max = 15.00)

def insert_missing(m) =
    if m == [] then
        [("title", "Live Broadcast"), ("is_live", "true")]
    else
        [("is_live", "true")]
    end
end
live = metadata.map(insert_missing, live)

radio = fallback(id="live_fallback", replay_metadata=true, [live, radio])

# AutoDJ Transition Settings
def live_aware_crossfade(old, new) =
    if !to_live then
        # If going to the live show, play a simple sequence
        sequence([fade.out(old.source),fade.in(new.source)])
    else
        # Otherwise, use the smart transition
        cross.simple(old.source, new.source, fade_in=3.00, fade_out=3.00)
    end
end

radio = cross(minimum=0., duration=4.50, live_aware_crossfade, radio)
```

If i place the `# AutoDJ Transition Settings` code part before `# DJ Authentication`, so i could add Stereo Tool in between, the behavior is changing: switching to live doesn't happen, or is happening immediatly, in the middle of AutoDJ song. 😔

**Comment by unudelaradio** [comment]
I also have this piece of code, which i deleted, so i can get that _**switch to live after current song**_ effect.

```
# Skip non-live track when live DJ goes live.
def check_live() =
    if live.is_ready() then
        if not to_live() then
            to_live.set(true)
            radio_without_live.skip()
        end
    else
        to_live.set(false)
    end
end

# Continuously check on live.
radio = source.on_frame(radio, check_live)
```

Maybe the secret is in this code? Should i change something, to get that **_waiting/deelay_** effect before switch?

**Comment by Moonbase59** [comment]
It’s an AzuraCast thing. Apparently, i older versions, the "live" came _before_ the crossfade, so it could be crossfaded as well. In theory, this would also allow "clean skips", i.e. no artifacts after live ends, because the skip could happen while we still have separate (unmixed) tracks.

Looking at AzuraCast’s code (LS ConfigWriter), it appears the "live" fallback has _intentionally_ been moved _after_ the crossfade, because some buffering problems were experienced. That, in turn, now brings _no_ crossfades and possibly the "fragment after live" problem.

Coincidence: I spent the whole Sunday trying to figure this out, but I won’t be able to come back to that before end of August.

**Reply by unudelaradio** [reply]
> because some buffering problems were experienced

Unfortunately yes!

It only works on older versions, with no buffering problems. 

My tests are like this:

Older Versions:
- _live_ then _crossfade_ (works perfect)
- _stereo tool_ then _live_ then _crossfade_ (chaotic behavior with short jingles in AutoDJ)

Newest Version:
-  _live_ then _crossfade_ (buffering problems, streamer is kicked out after few seconds)
- _crossfade_ then _live_ (no fades on switch and fragment of last AutoDJ song at the end of live)

Trying to decide which is more convenient.

**Reply by Moonbase59** [reply]
Same symptoms here. Let’s hope LS 2.3.x will bring some improvements on that.

I’d like playlists and live first, then be able to crossfade. Naturally _post-_ processing (master_me, Stereo Tool, etc.) always should be _last_ in the processing chain.
